File: lecture11/exercise11-B/solution.py
#!/usr/bin/env python3

# Exercise 11-B: Passcode Cracking

import collections
import sys

# Graph structure

Graph = collections.namedtuple('Graph', 'edges degrees')

# Read Graph

def read_graph() -> Graph:
    edges: dict[int, set[int]] = collections.defaultdict(set)
    for line in sys.stdin:
        digits = list(map(int, line.strip()))

        edges[digits[0]].add(digits[1])
        edges[digits[1]].add(digits[2])

    degrees: dict[int, int] = collections.defaultdict(int)
    for s, ts in edges.items():
        degrees[s]
        for t in ts:
            degrees[t] += 1

    return Graph(edges, degrees)

# Topological Sort

def topological_sort(graph: Graph) -> list[int]:
    frontier = [v for v, d in graph.degrees.items() if d == 0]
    visited  = []

    while frontier:
        vertex = frontier.pop()

        visited.append(vertex)

        for neighbor in graph.edges[vertex]:
            graph.degrees[neighbor] -= 1
            if graph.degrees[neighbor] == 0:
                frontier.append(neighbor)

    return visited

# Main Execution

def main() -> None:
    graph = read_graph()
    code  = topological_sort(graph)

    print(''.join(map(str, code)))

if __name__ == '__main__':
    main()

File: lecture11/exercise11-B/solution.cpp
// Exercise 11-B: Passcode Cracking

#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>

using namespace std;

// Structures

struct Graph {
    unordered_map<char, unordered_set<char>> edges;
    unordered_map<char, int>		     degrees;
};

// Functions

Graph read_graph() {
    Graph  graph;
    string digits;
    
    while (cin >> digits) {
    	graph.edges[digits[0]].insert(digits[1]);
    	graph.edges[digits[1]].insert(digits[2]);
    }

    for (auto e : graph.edges) {
    	auto s  = e.first;
    	auto ts = e.second;

    	graph.degrees[s];
    	for (auto t : ts) {
    	    graph.degrees[t] += 1;
	}
    }

    return graph;
}

string topological_sort(Graph &graph) {
    queue<int>	frontier;
    string	visited;

    for (auto d : graph.degrees) {
    	if (d.second == 0) {
    	    frontier.push(d.first);
	}
    }

    while (!frontier.empty()) {
    	auto vertex = frontier.front(); frontier.pop();

    	visited += vertex;

    	for (auto neighbor : graph.edges[vertex]) {
    	    graph.degrees[neighbor] -= 1;
    	    if (graph.degrees[neighbor] == 0) {
    	    	frontier.push(neighbor);
	    }
	}
    }

    return visited;
}

// Main Execution

int main(int argc, char *argv[]) {
    auto graph = read_graph();
    auto code  = topological_sort(graph);

    cout << code << endl;
    return EXIT_SUCCESS;
}
